### PR TITLE
Additional linux AF_PACKET functionality

### PR DESCRIPTION
**Pull request by nathaniel-bennett**
Added socket options, structs and other miscellaneous constants that make memory-mapped packet sockets possible in Linux (PACKET_TX_RING, PACKET_RX_RING, etc.).

### CODE REVIEW

**Comment by nathaniel-bennett** [comment]
(Note that this change removes the `netpacket/packet.h` header file and replaces it with `linux/if_packet.h`. This is because if_packet.h is a proper superset of packet.h, and the inclusion of both would lead to build errors with structs/constants being defined twice)

**Review by JohnTitor — APPROVED** [review]

**Comment by JohnTitor** [comment]
The failure is valid, I guess the cause is our musl image on CI is old, we could skip tests for that arch.

**Comment by nathaniel-bennett** [comment]
It seems like `fanout_args` was the main culprit  that was causing the musl build to fail, so I've added a check that skips that particular struct for musl tests.

**Comment by nathaniel-bennett** [comment]
Alright, so this time the musl build failed due to some issue with `sockaddr_pkt` (likely something to do with endianness). I've opted to just remove `sockaddr_pkt`, since it's considered obsolete. It was part of the original `SOCK_PACKET` raw socket interface for Linux that was superceded by `AF_PACKET` and the accompanying `sockaddr_ll` socket type many years ago.

Fingers crossed that this time we actually pass the extended checks :)

**Comment by JohnTitor** [comment]
@bors r-
Could you remove commented out `sockaddr_pkt`?

File: libc-test/build.rs [hunk 1]
```diff
@@ -3377,6 +3377,8 @@ fn test_linux(target: &str) {
             // FIXME: This is actually a union, not a struct
             "sigval" => true,
 
+            "fanout_args" if musl => true,
```
Comment (JohnTitor) [inline libc-test/build.rs:3380]: A comment why this is ignored is helpful when we upgrade CI toolchains :)

**Comment by workingjubilee** [comment]
Oh, apologies. I checked twice for the struct definition and thought the latest force-push had fully removed it.

**Comment by nathaniel-bennett** [comment]
Removed commented out code and added comment on why "fanout_args" was ignored :+1:

**Comment by JohnTitor** [comment]
Thank you both! @bors r+

**Comment by JohnTitor** [comment]
Ah, sparc64's image is also old and we have to ignore tests for that too :(

**Comment by nathaniel-bennett** [comment]
Added exception for sparc64 (with a comment on removing it once CI is upgraded)

**Comment by JohnTitor** [comment]
Great, thanks a lot! Let's try again:
@bors r+

**Comment by nathaniel-bennett** [comment]
This most recent build error feels like some problem with endianness and padding:

```
rust [16] = 239 != 0 (C): C "tpacket_auxdata" -> Rust
rust [17] = 238 != 0 (C): C "tpacket_auxdata" -> Rust
rust [18] = 237 != 0 (C): C "tpacket_auxdata" -> Rust
rust [19] = 236 != 0 (C): C "tpacket_auxdata" -> Rust
rust[8] = 0 != 8 (C): Rust "tpacket_hdr_variant1" -> C
rust[9] = 0 != 9 (C): Rust "tpacket_hdr_variant1" -> C
rust[10] = 0 != 10 (C): Rust "tpacket_hdr_variant1" -> C
rust[11] = 0 != 11 (C): Rust "tpacket_hdr_variant1" -> C
rust [24] = 231 != 0 (C): C "tpacket_req3" -> Rust
rust [25] = 230 != 0 (C): C "tpacket_req3" -> Rust
rust [26] = 229 != 0 (C): C "tpacket_req3" -> Rust
rust [27] = 228 != 0 (C): C "tpacket_req3" -> Rust
rust[8] = 0 != 8 (C): Rust "tpacket_stats_v3" -> C
rust[9] = 0 != 9 (C): Rust "tpacket_stats_v3" -> C
rust[10] = 0 != 10 (C): Rust "tpacket_stats_v3" -> C
rust[11] = 0 != 11 (C): Rust "tpacket_stats_v3" -> C
rust [24] = 231 != 0 (C): C "tpacket_req_u" -> Rust
rust [25] = 230 != 0 (C): C "tpacket_req_u" -> Rust
rust [26] = 229 != 0 (C): C "tpacket_req_u" -> Rust
rust [27] = 228 != 0 (C): C "tpacket_req_u" -> Rust
```

This is happening on `sparc64`, which is both 64-bit *and* big-endian. It looks like any struct that doesn't end on a 64-bit boundary fails the test because the last 4-byte field (or two 2-byte fields) are zeros when they should have values/vice versa. Maybe the big enndian architecture is padding to 8 bytes, and moving the last field in the struct to the last 4 bytes?

What *really* confuses me is why this isn't happening to other struct definitions that are being tested that have the same property (like `uinput_ff_erase`).

**Comment by nathaniel-bennett** [comment]
Would this be fixed by adding something like repr(align(8))? Or maybe adding a 4-byte `__pad` field that only exists when the architecture is big-endian and 64-bit?

**Comment by nathaniel-bennett** [comment]
The plot thickens--looks like `uinput_ff_erase` is actually skipped by checks:
```
        // FIXME: musl CI has old headers
        if (musl || sparc64) && ty.starts_with("uinput_") {
            return true;
        }
```
So this is a problem that has been encountered in the past.

 It seems that the `musl || sparc64` combo happens to cover all architectures that are both 64-bit *and* big-endian--which would point toward it being an endianness + padding issue.

**Comment by nathaniel-bennett** [comment]
I've opted to have the failing `sparc64` tests be explicitly skipped by CI (like what other structs do), but I'm going to open a new pull request that resolves the more widespread issue of big-endian padding :thumbsup:. If my hypothesis is correct, this is what is causing #1558 as well.

**Comment by nathaniel-bennett** [comment]
Never mind--I did a bit more digging and there isn't any padding in these structs to mix up. #1558 covers the same kind of issue; my guess is that once it is resolved, this will be also.

**Comment by JohnTitor** [comment]
Sorry for the delay!
@bors r+

**Comment by JohnTitor** [comment]
Failed on mips: @rustbot author

**Comment by nathaniel-bennett** [comment]
Did a rebase for this--the issues with mips and BSD should (hopefully) all be fixed now

**Comment by nathaniel-bennett** [comment]
Most recent issue was with MIPS--it's in the same camp as sparc64 and musl with big endian alignment. I've added an exception for `sockaddr_pkt` for mips, similar to what I did for the others.

```text
     Running test/main.rs (target/mips-unknown-linux-gnu/debug/deps/main-ae76ee3ebd17d880)
RUNNING ALL TESTS
rust[16] = 0 != 16 (C): Rust "sockaddr_pkt" -> C
rust[17] = 0 != 17 (C): Rust "sockaddr_pkt" -> C
```

**Comment by JohnTitor** [comment]
Seems something is wrong on powerpc64:
```
 bad TPACKET3_HDRLEN value at byte 7: rust: 84 (0x54) != c 68 (0x44)
bad tpacket_auxdata size: rust: 24 (0x18) != c 20 (0x14)
bad field offset tp_vlan_tci of tpacket_auxdata: rust: 20 (0x14) != c 16 (0x10)
bad field type tp_vlan_tci of tpacket_auxdata: rust: 0x40027409dc != c 0x40027409d8
bad field offset tp_vlan_tpid of tpacket_auxdata: rust: 22 (0x16) != c 18 (0x12)
bad field type tp_vlan_tpid of tpacket_auxdata: rust: 0x4002740a2e != c 0x4002740a2a
size of struct tpacket_auxdata is 20 in C and 24 in Rust
bad tpacket_hdr_variant1 size: rust: 16 (0x10) != c 12 (0xc)
bad field offset tp_vlan_tpid of tpacket_hdr_variant1: rust: 12 (0xc) != c 8 (0x8)
bad field type tp_vlan_tpid of tpacket_hdr_variant1: rust: 0x4002740ae4 != c 0x4002740ae0
bad field offset tp_padding of tpacket_hdr_variant1: rust: 14 (0xe) != c 10 (0xa)
bad field type tp_padding of tpacket_hdr_variant1: rust: 0x4002740b1e != c 0x4002740b1a
size of struct tpacket_hdr_variant1 is 12 in C and 16 in Rust
bad tpacket_req3 size: rust: 32 (0x20) != c 28 (0x1c)
bad field offset tp_feature_req_word of tpacket_req3: rust: 28 (0x1c) != c 24 (0x18)
bad field type tp_feature_req_word of tpacket_req3: rust: 0x4002740a2c != c 0x4002740a28
size of struct tpacket_req3 is 28 in C and 32 in Rust
bad tpacket_stats_v3 size: rust: 16 (0x10) != c 12 (0xc)
bad field offset tp_freeze_q_cnt of tpacket_stats_v3: rust: 12 (0xc) != c 8 (0x8)
bad field type tp_freeze_q_cnt of tpacket_stats_v3: rust: 0x4002740b6c != c 0x4002740b68
size of struct tpacket_stats_v3 is 12 in C and 16 in Rust
bad tpacket3_hdr size: rust: 52 (0x34) != c 48 (0x30)
bad field offset tp_padding of tpacket3_hdr: rust: 44 (0x2c) != c 40 (0x28)
bad field type tp_padding of tpacket3_hdr: rust: 0x400274097c != c 0x4002740978
size of struct tpacket3_hdr is 48 in C and 52 in Rust
```
@rustbot author

**Comment by JohnTitor** [comment]
Closing as inactive, feel free to re-submit a PR if you're still interested in this change. Thanks anyway!

**Comment by nathaniel-bennett** [comment]
I kept running into struct alignment issues on 64-bit big-endian architectures (currently being tracked at https://github.com/rust-lang/rust/issues/43894#issuecomment-1768787736); these would lead to bors failing. Once rust-lang resolves those, I'll resubmit a PR and get this functionality merged in.

**Pull request closed without merging**